services/httpd/handler.go

```diff
@@ -993,8 +1017,437 @@ func fixLiterals(node influxql.Node) {
+		db, rp, err = bucket2dbrp(after)
+		if err != nil {
+			h.httpError(w, fmt.Sprintf("list buckets invalid parameter - after=%q: %s", after, err.Error()), http.StatusNotFound)
+			return
+		}
+	}
+
+	if limitStr := r.URL.Query().Get("limit"); limitStr != "" {
+		if l, err := strconv.ParseInt(limitStr, 10, 64); err != nil {
+			h.httpError(w, fmt.Sprintf("list buckets parameter is not an integer - limit=%q: %s", limitStr, err.Error()), http.StatusBadRequest)
+			return
+		} else if l <= 0 {
+			limit = math.MaxInt
+		} else {
+			limit = int(l)
+		}
+	}
+
+	if offsetStr := r.URL.Query().Get("offset"); offsetStr != "" {
+		if after != "" {
+			h.httpError(w, "list buckets cannot have both \"offset\" and \"after\" arguments", http.StatusBadRequest)
+			return
+		} else if o, err := strconv.ParseInt(offsetStr, 10, 64); err != nil {
+			h.httpError(w, fmt.Sprintf("list buckets parameter is not an integer - offset=%q: %s", offsetStr, err.Error()), http.StatusBadRequest)
+			return
+		} else {
+			offset = int(o)
+		}
+	}
+
+	dbIndex := 0
+	rpIndex := 0
+	dbs := h.MetaClient.Databases()
+	if after != "" {
+		if dbIndex, rpIndex = findBucketIndex(dbs, db, rp); dbIndex < 0 {
+			h.httpError(w, fmt.Sprintf("list buckets \"after\" parameter not found: %q", after), http.StatusNotFound)
+			return
+		}
+	} else if offset > 0 {
+		if dbIndex, rpIndex = findBucketOffsetIndex(dbs, offset); dbIndex < 0 {
+			sendBuckets(w, []Bucket{})
+			return
+		}
+	}
+	sliceCap := limit
+	if sliceCap > (2 * defaultLimit) {
+		sliceCap = 2 * defaultLimit
+	}
+	buckets := make([]Bucket, 0, sliceCap)
+
+outer:
+	for ; dbIndex < len(dbs); dbIndex++ {
+		dbi := dbs[dbIndex]
+		if h.Config.AuthEnabled {
+			if err := h.QueryAuthorizer.AuthorizeDatabase(user, influxql.ReadPrivilege, dbi.Name); err != nil {
+				continue
+			}
+		}
+		for ; rpIndex < len(dbi.RetentionPolicies); rpIndex++ {
+			if limit > 0 {
+				buckets = append(buckets, *makeBucket(&dbi.RetentionPolicies[rpIndex], dbi.Name))
+				limit--
+			} else {
+				break outer
+			}
+		}
+		rpIndex = 0
+	}
+	sendBuckets(w, buckets)
+}
+
+func (h *Handler) oneBucket(w http.ResponseWriter, name string) *Bucket {
+	db, rp, err := bucket2dbrp(name)
+	if err != nil {
+		h.httpError(w, fmt.Sprintf("bucket %q: %s", name, err.Error()), http.StatusNotFound)
+		return nil
+	} else if dbi := h.MetaClient.Database(db); dbi == nil {
+		h.httpError(w, fmt.Sprintf("bucket not found: %q", name), http.StatusNotFound)
+		return nil
+	} else if rpi := dbi.RetentionPolicy(rp); rpi == nil {
+		h.httpError(w, fmt.Sprintf("bucket not found: %q", name), http.StatusNotFound)
+		return nil
+	} else {
+		return makeBucket(rpi, db)
+	}
+}
+
+func findBucketOffsetIndex(dbs []meta.DatabaseInfo, offset int) (dbIndex int, rpIndex int) {
+	for i, dbi := range dbs {
+		if offset > len(dbi.RetentionPolicies) {
+			offset -= len(dbi.RetentionPolicies)
+		} else {
+			dbIndex = i
+			rpIndex = offset
+			return dbIndex, rpIndex
+		}
+	}
+	return -1, -1
```

Comment:
I think I'd prefer a bool `ok` parameter that says found / not found, rather than the negative index. Similar to `ok` for channels, maps, etc.